File: Day1-9/4.py
def part1_valid(pw):
    words = pw.split()
    return len(set(words)) == len(words)


def part2_valid(pw):
    words = pw.split()
    sorted_words = [''.join(sorted(w)) for w in words]
    return len(set(sorted_words)) == len(words)


def count_valid_passwords(lines, validator):
    return sum(1 for pw in lines if validator(pw))


if __name__ == '__main__':
    with open('4.txt') as f:
        lines = f.readlines()
    print("Part 1: {}".format(
        count_valid_passwords(lines, part1_valid))
    )
    print("Part 2: {}".format(
        count_valid_passwords(lines, part2_valid))
    )

File: Day1-9/4.go
package main

import (
	"fmt"
	"io/ioutil"
	"sort"
	"strings"
)

type validator func(string) bool

func part1Validator(input string) bool {
	words := strings.Split(input, " ")
	wordSet := make(map[string]bool)
	for _, w := range words {
		wordSet[w] = true
	}
	return len(wordSet) == len(words)
}

func part2Validator(input string) bool {
	words := strings.Split(input, " ")
	wordSet := make(map[string]bool)
	for _, w := range words {
		s := strings.Split(w, "")
		sort.Strings(s)
		wordSet[strings.Join(s, "")] = true
	}
	return len(wordSet) == len(words)
}

func countValidPasswords(candidates []string, isValid validator) int {
	count := 0
	for _, pw := range candidates {
		if isValid(pw) {
			count++
		}
	}
	return count
}

func main() {
	dat, _ := ioutil.ReadFile("./4.txt")
	lines := strings.Split(string(dat), "\n")
	fmt.Printf("Part 1: %d\n", countValidPasswords(lines, part1Validator))
	fmt.Printf("Part 2: %d\n", countValidPasswords(lines, part2Validator))
}
